tests/simulator_test.cpp

```diff
@@ -2,82 +2,284 @@
-  ASSERT_FALSE(performHitAtEnergyCollectors(tiltedFromXY, &hitData));
-
-  Ray straightUpMoved(
-      Vec3(2 * manager.collectorRadius(), 2 * manager.collectorRadius(), 0),
-      Vec3(0, 0, 1));
-  ASSERT_FALSE(performHitAtEnergyCollectors(straightUpMoved, &hitData));
-} */
\ No newline at end of file
+  ASSERT_TRUE(performHitCollector(energyCollectors, straightUp, &hitData));
+
+  const float collectorPositionRadius = 4;
+  const float collectorAngle = 2 * kPi / (numCollectors - 1);
+  const float refCollectorRadius =
+      collectorPositionRadius * std::sqrt(2 - 2 * std::cos(collectorAngle));
+  Vec3 collisionPointStraightUp =
+      Vec3(0, 0, collectorPositionRadius - refCollectorRadius);
+  ASSERT_EQ(hitData.collisionPoint(), collisionPointStraightUp);
+
+  Ray straightDown(kVecZero, -kVecUp);
+  ASSERT_FALSE(performHitCollector(energyCollectors, straightDown, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+
+  Ray alongX(kVecZero, kVecX);
+  ASSERT_TRUE(performHitCollector(energyCollectors, alongX, &hitData));
+
+  ASSERT_EQ(collectorPositionRadius - refCollectorRadius, hitData.time);
+
+  Ray alongY(kVecZero, kVecY);
+  ASSERT_TRUE(performHitCollector(energyCollectors, alongY, &hitData));
+
+  ASSERT_EQ(alongY.at(collectorPositionRadius - refCollectorRadius),
+            hitData.collisionPoint());
+
+  Ray atSixty(kVecZero, Vec3(std::cos(deg2rad(60)), 0, std::sin(deg2rad(60))));
+  ASSERT_TRUE(performHitCollector(energyCollectors, atSixty, &hitData));
+
+  Ray atSixtyOther(kVecZero,
+                   Vec3(-std::cos(deg2rad(60)), 0, std::sin(deg2rad(60))));
+  ASSERT_TRUE(performHitCollector(energyCollectors, atSixtyOther, &hitData));
+
+  Ray atSixtyXY(kVecZero, Vec3(std::cos(deg2rad(60)), std::cos(deg2rad(60)),
+                               std::sin(deg2rad(60))));
+  ASSERT_FALSE(performHitCollector(energyCollectors, atSixtyXY, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+
+  Ray atSixtyXYOther(kVecZero,
+                     Vec3(-std::cos(deg2rad(60)), -std::cos(deg2rad(60)),
+                          std::sin(deg2rad(60))));
+  ASSERT_FALSE(performHitCollector(energyCollectors, atSixtyXYOther, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+}
+
+TEST_F(EnergyCollectorTest, EvenNumOfEnergyCollectorTest) {
+  const FakeModel nonEmptyModel(false);
+
+  const int numCollectors = 36;
+  auto energyCollectors = buildCollectors(nonEmptyModel, numCollectors);
+  ASSERT_EQ(energyCollectors.size(), numCollectors);
+
+  Ray straightUp(kVecZero, kVecUp);
+  RayHitData hitData;
+  ASSERT_TRUE(performHitCollector(energyCollectors, straightUp, &hitData));
+
+  const float collectorPositionRadius = 4;
+  const float collectorAngle = 2 * kPi / (numCollectors - 2);
+  const float refCollectorRadius =
+      collectorPositionRadius * std::sqrt(2 - 2 * std::cos(collectorAngle));
+
+  std::pair<objects::EnergyCollector, objects::EnergyCollector> topCol =
```

Comment:
Why do you need both of them? Aren't they on the same Z level, and (X1+X2)/2 = 0 ?

Can't you just return Z as we discussed yesterday? if they are not on the same Z level, maybe code is wrong?

Reply:
done